File: wagerproof-ios-native/Wagerproof/Features/MLB/Components/MLBGameCard.swift
import SwiftUI
import WagerproofDesign
import WagerproofModels

/// MLB game row rendered in the home Games feed list. MLB publishes
/// `homeMlEdgePct` / `awayMlEdgePct` directly so we don't have to
/// recompute the ML edge from probabilities, and `ouFairTotal` for the
/// O/U delta.
struct MLBGameCard: View {
    let game: MLBGame
    var onPress: () -> Void = {}

    var body: some View {
        GameRowCard(model: rowModel, onPress: onPress)
    }

    private var rowModel: GameRowCard.Model {
        let awayDisplayName = game.awayTeamName ?? game.awayTeam ?? "Away"
        let homeDisplayName = game.homeTeamName ?? game.homeTeam ?? "Home"

        let awayPair = MLBTeams.colors(for: awayDisplayName)
        let homePair = MLBTeams.colors(for: homeDisplayName)

        return GameRowCard.Model(
            id: game.id,
            league: "mlb",
            dateLabel: MLBFormatting.dateLabel(game.officialDate),
            timeLabel: MLBFormatting.gameTime(game.gameTimeEt),
            away: GameRowCard.TeamSide(
                abbr: game.awayAbbr,
                initials: game.awayAbbr,
                moneyline: game.awayMl,
                spread: game.awaySpread,
                logoURL: game.awayLogoUrl,
                colors: TeamColorPair(
                    primary: Color(hex: Int(awayPair.primary)),
                    secondary: Color(hex: Int(awayPair.secondary))
                )
            ),
            home: GameRowCard.TeamSide(
                abbr: game.homeAbbr,
                initials: game.homeAbbr,
                moneyline: game.homeMl,
                spread: game.homeSpread,
                logoURL: game.homeLogoUrl,
                colors: TeamColorPair(
                    primary: Color(hex: Int(homePair.primary)),
                    secondary: Color(hex: Int(homePair.secondary))
                )
            ),
            overLine: game.totalLine,
            mlEdge: mlbMlEdge,
            ouEdge: mlbOuEdge,
            awayTeamFullName: awayDisplayName,
            homeTeamFullName: homeDisplayName,
            oddsBreakdown: oddsBreakdown
        )
    }

    /// Build the Spread / Money / Total table for the row. MLB only stores
    /// the run line, moneyline, and a shared total (no spread/total juice),
    /// so each cell shows the line we have. Total uses the Over/Under
    /// convention: Over on the away row, Under on the home row.
    private var oddsBreakdown: GameRowCard.OddsBreakdown {
        let totalText = MLBFormatting.line(game.totalLine)
        return GameRowCard.OddsBreakdown(
            away: GameRowCard.OddsBreakdown.Row(
                abbr: game.awayAbbr,
                spread: MLBFormatting.spread(game.awaySpread),
                moneyline: MLBFormatting.moneyline(game.awayMl),
                total: game.totalLine != nil ? "O\(totalText)" : "—"
            ),
            home: GameRowCard.OddsBreakdown.Row(
                abbr: game.homeAbbr,
                spread: MLBFormatting.spread(game.homeSpread),
                moneyline: MLBFormatting.moneyline(game.homeMl),
                total: game.totalLine != nil ? "U\(totalText)" : "—"
            )
        )
    }

    /// Prefer the MLB predictions' published `*MlEdgePct` columns over
    /// recomputing edge from probabilities — the backend already does
    /// the line-shopping math we'd be redoing.
    private var mlbMlEdge: GameRowCard.MLEdgeInfo? {
        let homeEdge = game.homeMlEdgePct
        let awayEdge = game.awayMlEdgePct
        if let homeEdge, let awayEdge {
            if homeEdge >= awayEdge {
                return GameRowCard.MLEdgeInfo(
                    abbr: game.homeAbbr,
                    edgePoints: homeEdge,
                    color: GameEdgeMath.edgeColor(homeEdge)
                )
            } else {
                return GameRowCard.MLEdgeInfo(
                    abbr: game.awayAbbr,
                    edgePoints: awayEdge,
                    color: GameEdgeMath.edgeColor(awayEdge)
                )
            }
        }
        // Fall back to the model probability calc if edges are missing.
        return GameEdgeMath.mlEdge(
            modelHomeProb: game.mlHomeWinProb,
            homeMl: game.homeMl,
            awayMl: game.awayMl,
            homeAbbr: game.homeAbbr,
            awayAbbr: game.awayAbbr
        )
    }

    /// MLB has `ouDirection` ("OVER"/"UNDER") + `ouEdge` (signed pct
    /// points) + `ouFairTotal`. Use the published direction directly when
    /// present so it matches the back-end's bet labeling.
    private var mlbOuEdge: GameRowCard.OUEdgeInfo? {
        if let direction = game.ouDirection {
            let isOver = direction.uppercased() == "OVER"
            let delta: Double? = {
                guard let fair = game.ouFairTotal, let line = game.totalLine else { return nil }
                return fair - line
            }()
            let edgeMag = abs(game.ouEdge ?? 0)
            return GameRowCard.OUEdgeInfo(
                isOver: isOver,
                delta: delta,
                probability: nil,
                color: GameEdgeMath.edgeColor(edgeMag)
            )
        }
        return GameEdgeMath.ouEdge(
            modelFairTotal: game.ouFairTotal,
            marketLine: game.totalLine,
            ouResultProb: nil
        )
    }
}

File: wagerproof-android-native/app/src/main/java/com/wagerproof/app/features/mlb/MLBGameCard.kt
package com.wagerproof.app.features.mlb

import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import com.wagerproof.app.features.gamecards.GameEdgeMath
import com.wagerproof.app.features.gamecards.GameRowCard
import com.wagerproof.app.features.gamecards.GameRowCardModel
import com.wagerproof.core.models.MLBGame
import kotlin.math.abs

/**
 * MLB game row for the home Games feed — a thin adapter over the shared
 * [GameRowCard], mirroring iOS `MLBGameCard`. MLB publishes `homeMlEdgePct` /
 * `awayMlEdgePct` + `ouDirection`/`ouFairTotal`, so we prefer those published
 * columns over recomputing edge from probabilities (falling back to
 * [GameEdgeMath] when a game hasn't been scored yet).
 */
@Composable
fun MLBGameCard(
    game: MLBGame,
    onPress: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val awayDisplayName = game.awayTeamName ?: game.awayTeam ?: "Away"
    val homeDisplayName = game.homeTeamName ?: game.homeTeam ?: "Home"

    val model = GameRowCardModel(
        id = game.id,
        league = "mlb",
        dateLabel = MLBFormatting.dateLabel(game.officialDate),
        timeLabel = MLBFormatting.gameTime(game.gameTimeEt),
        away = GameRowCardModel.TeamSide(
            abbr = game.awayAbbr,
            initials = game.awayAbbr,
            moneyline = game.awayMl,
            spread = game.awaySpread,
            logoURL = game.awayLogoUrl,
            colors = mlbTeamColorPair(awayDisplayName),
        ),
        home = GameRowCardModel.TeamSide(
            abbr = game.homeAbbr,
            initials = game.homeAbbr,
            moneyline = game.homeMl,
            spread = game.homeSpread,
            logoURL = game.homeLogoUrl,
            colors = mlbTeamColorPair(homeDisplayName),
        ),
        overLine = game.totalLine,
        mlEdge = mlbMlEdge(game),
        ouEdge = mlbOuEdge(game),
        awayTeamFullName = awayDisplayName,
        homeTeamFullName = homeDisplayName,
        oddsBreakdown = oddsBreakdown(game),
    )

    GameRowCard(model = model, onPress = onPress, modifier = modifier)
}

/** Run line / ML / shared total table. Total uses O on the away row, U on home. */
private fun oddsBreakdown(game: MLBGame): GameRowCardModel.OddsBreakdown {
    val totalText = MLBFormatting.line(game.totalLine)
    val hasTotal = game.totalLine != null
    return GameRowCardModel.OddsBreakdown(
        awaySpread = MLBFormatting.spread(game.awaySpread),
        homeSpread = MLBFormatting.spread(game.homeSpread),
        awayML = MLBFormatting.moneyline(game.awayMl),
        homeML = MLBFormatting.moneyline(game.homeMl),
        awayTotal = if (hasTotal) "O$totalText" else "—",
        homeTotal = if (hasTotal) "U$totalText" else "—",
    )
}

/** Prefer the published `*MlEdgePct` columns; fall back to the probability calc. */
private fun mlbMlEdge(game: MLBGame): GameRowCardModel.MLEdgeInfo? {
    val homeEdge = game.homeMlEdgePct
    val awayEdge = game.awayMlEdgePct
    if (homeEdge != null && awayEdge != null) {
        return if (homeEdge >= awayEdge) {
            GameRowCardModel.MLEdgeInfo(game.homeAbbr, homeEdge, GameEdgeMath.edgeColor(homeEdge))
        } else {
            GameRowCardModel.MLEdgeInfo(game.awayAbbr, awayEdge, GameEdgeMath.edgeColor(awayEdge))
        }
    }
    return GameEdgeMath.mlEdge(
        modelHomeProb = game.mlHomeWinProb,
        homeMl = game.homeMl,
        awayMl = game.awayMl,
        homeAbbr = game.homeAbbr,
        awayAbbr = game.awayAbbr,
    )
}

/** Prefer the published `ouDirection`/`ouFairTotal`; fall back to fair-vs-market. */
private fun mlbOuEdge(game: MLBGame): GameRowCardModel.OUEdgeInfo? {
    val direction = game.ouDirection
    if (direction != null) {
        val isOver = direction.uppercase() == "OVER"
        val fair = game.ouFairTotal
        val line = game.totalLine
        val delta = if (fair != null && line != null) fair - line else null
        val edgeMag = abs(game.ouEdge ?: 0.0)
        return GameRowCardModel.OUEdgeInfo(
            isOver = isOver,
            delta = delta,
            // MLB does not publish a directional confidence here. Showing a
            // fabricated 50% made every Android row longer and contradicted
            // the iOS card, which intentionally omits this value.
            probability = null,
            color = GameEdgeMath.edgeColor(edgeMag),
        )
    }
    return GameEdgeMath.ouEdge(
        modelFairTotal = game.ouFairTotal,
        marketLine = game.totalLine,
        ouResultProb = null,
    )
}
